### ISSUE TITLE
Enzymes not showed

### ISSUE THREAD

**Issue by Sanrrone** [Lattice-Automation/seqviz#88]
Dears,
I tried to see in the most basic example the enzymes "PstI" and "EcoRI"  using the following code:

```
<script>
  window.seqviz
    .Viewer("test", {
      name: "test",
      seq: "tcgcgcgtttcggtCATGCACTGCAGGTCAGgatgacggtgaaaacctctgacaCATGCACTGCAGGTCAGtcagct",
      style: { height: "100vh", width: "100vw" },
      enzymes: ["PstI", "EcoRI"]
    })
    .render();
</script>
```
But only the basic funtionality works, I can see the sequence as circular as linear. I can select, but nothing happen wihen I try the enzymes options.
 I'm using the link `<script src="https://unpkg.com/seqviz"></script>` and a local copy but none of them works for now.

is there other way to see an enzyme in my sequence?
thanks in advance,
Regards

**Comment by Sanrrone** [comment]
As additional comment, if I replace name and seq tags by `file` (with a gb file), it works. But I'm looking to put the sequence directly in the seq tag.

**Comment by jjti** [comment]
you're right, it was bugged and not cutting if a `seq` was provided. It should be fixed in v3.0.27: https://www.npmjs.com/package/seqviz

feel free to re-open if I'm wrong and issue persists, but I believe it's now fixed (please ignore the ugly enzymes lines on the smallish circular plasmid)

<img width="1344" alt="Screen Shot 2021-02-13 at 6 01 34 PM" src="https://user-images.githubusercontent.com/13923102/107864235-154c4800-6e28-11eb-905f-01437f6546b3.png">

**Comment by Sanrrone** [comment]
Hi,
I found another bug related to this. if I create my own enzyme (just to test) called "Cas9" with a NGG seq of recognition, Its stop to find matches after second match.
![issue](https://user-images.githubusercontent.com/15197238/107951686-7898c000-6f77-11eb-9b7c-c83ab7d7eba9.png)

also, in this case the sequence NGG, only the GGs are visually wrapped (not really important by the way).

thanks for the previous fix!

**Comment by jjti** [comment]
do you mind posting the example seq and recognition seq so I can repoduce? Also, it appears that you added the enzyme to the `utils/enzymes.js` module, is that right?

**Comment by Sanrrone** [comment]
Sure, I created a minimal example to reproduce it.
```
<script>
  window.seqviz
    .Viewer("root", {
      name: "L09136",
      seq: "AAAAAAAAAATGGAAAAAAAAAGGAAAAAAAAAAANGGAAAAAAAATTGGTTTTTTTTTGGTTTT",
      enzymes: ["Cas9"],
      style: { height: "100vh", width: "100vw" }
    })
    .render();
</script>
```
regarding to the enzyme, I'm not using the modules but I added the enzyme directly in the js (which is the same I guess for the enzymes.js)
eg: `"Cas9":{rseq:"NGG",rcut:0,fcut:0}`

Best,
Sandro

**Comment by Mokey2002** [comment]
@Sanrrone 
Hi, quick question.
 How were you able to add custom enzymes using <script src="https://unpkg.com/seqviz"></script>? 

Thanks,
Eduardo

**Comment by Sanrrone** [comment]
Hi @Mokey2002 ,
I just downloaded the entire js and modify it in the enzyme section.

Saludos,
Sandro

**Comment by Mokey2002** [comment]
Hi @Sanrrone,

Muchas gracias

Saludos,
Eduardo

**Comment by jjti** [comment]
I just pushed a version `3.0.29` where this should be fixed. I wouldn't be surprised if there are still issues with digest, but it's now less-wrong AFAIK. Sorry this is piecemeal, but let me know if you see something else, I didn't have time to really set up better testing for digest/enzymes

<img width="1357" alt="Screen Shot 2021-02-17 at 10 50 20 PM" src="https://user-images.githubusercontent.com/13923102/108303173-6e431580-7173-11eb-89da-8987cd1c6a26.png">

Also I added an option for custom enzymes:

### `options.enzymesCustom (={})`

Unsupported enzymes can also be passed through an object where the keys are the enzymes' names and the values are the enzymes:

```js
{
  "Cas9": {
    rseq: "NGG", // recognition sequence
    fcut: 0, // cut index on FWD strand, relative to start of rseq
    rcut: 1, // cut index on REV strand, relative to start of rseq
  }
}
```

**Comment by Sanrrone** [comment]
I used a minimum example
```
<script src="https://unpkg.com/seqviz"></script>
<script>
  window.seqviz
    .Viewer("root", {
      name: "L09136",
      seq: "TTATGTATGCGTTGTCCTTGGAGTATTAATATTGTTCATGTGGGCAGGCTCAGGTTGAGGTTGAGGTTGAGGGAACTGCTGTTCCTGT",
      enzymesCustom: {
          "Cas9": {
            rseq: "NGG", // recognition sequence
            fcut: 0, // cut index on FWD strand, relative to start of rseq
            rcut: 1, // cut index on REV strand, relative to start of rseq
          }]
        },
      style: { height: "100vh", width: "100vw" },
      search: { "query": "TTAATATT", "mismatch": 1 }
    })
    .render();
</script>
```
but it looks like enzymesCustom option is no available yet.
Additionally, I tried to `search` a query inside but nothing happens neither.

![Captura de pantalla de 2021-02-18 09-50-36](https://user-images.githubusercontent.com/15197238/108359978-63b66980-71cf-11eb-8c8e-ec005ebd1ce5.png)

**Comment by jjti** [comment]
when you get a chance, please try again with the latest version: https://unpkg.com/seqviz@3.0.30/dist/seqviz.min.js

also in your example there's an extra `]` bracket at the end of the custom enzymes, I made the below with

```
const viewerInput = {
  // name: "L09136",
  // seq: "tcgcgcgtttcggtgatgacggtgaaaacctctgacacatgca",
  seq: "TTATGTATGCGTTGTCCTTGGAGTATTAATATTGTTCATGTGGGCAGGCTCAGGTTGAGGTTGAGGTTGAGGGAACTGCTGTTCCTGT",
  enzymesCustom: {
    "Cas9": {
      rseq: "NGG", // recognition sequence
      fcut: 0, // cut index on FWD strand, relative to start of rseq
      rcut: 1, // cut index on REV strand, relative to start of rseq
    }
  },
  viewer: "both",
  // accession: "NC_011521",
  // backbone: "pSB1C3",
  showAnnotations: true,
  showPrimers: true,
  showComplement: true,
  showIndex: true,
  zoom: { linear: 50 },
  colors: [],
  onSearch: results => {
    console.log("Your Search Results: ", results);
  },
  bpColors: {
    10: "green",
    200: "blue",
    201: "red"
  },
  search: { query: "gtacc" },
  enzymes: ["EcoRI"],
  copyEvent: event => event.key === "c" && (event.metaKey || event.ctrlKey),
  style: { height: "calc(100vh - 20px)", width: "calc(100vw)" }
};
```

<img width="1352" alt="Screen Shot 2021-02-18 at 6 41 37 PM" src="https://user-images.githubusercontent.com/13923102/108436166-f16a7700-7218-11eb-8fef-c77673134c7e.png">

**Comment by Sanrrone** [comment]
Everything works fine now. 
Thanks for your help!

**Comment by Sanrrone** [comment]
Dear,
I found another little bug (but easy to fix I think), if I only use `enzymesCustom` there is no enzymes in my plot. But, if I use `enzymes` option + `enzymesCustom` all the enzymes are now showed.

the test case I used to find the bug:
```
<script>
  window.seqviz
    .Viewer("root", {
      seq: "TTATGTATGGCGTTGTCCTTGGAGTATTAATATTGTTCATGTGGGCAGGCTCAGGT",
      enzymes: ["psTI"], //comment this line to activate the bug
      enzymesCustom: {
          "TestEC": {
            rseq: "NGG", // recognition sequence
            fcut: 0, // cut index on FWD strand, relative to start of rseq
            rcut: 1, // cut index on REV strand, relative to start of rseq
          },
      },
      style: { height: "100vh", width: "100vw" },
    })
    .render();
</script>
```

In other subject, is there a way to set a windows of search? for example.
```
enzymesCustom: {
          "TestEC": {
            rseq: "NGG", // recognition sequence
            fcut: 0, // cut index on FWD strand, relative to start of rseq
            rcut: 1, // cut index on REV strand, relative to start of rseq
            wsearch: [ "5:15", "20:30" ] //meaning the range for different windows searchs for NGG pattern
          },
}
```
in that way the enzymesCustom "TestEC" would displayed only between position 5 to 15 and 20 to 30 (if there is a match), no wsearch would mean search in whole sequence. Is that possible?

thanks you in advance!
Regards

**Comment by jjti** [comment]
You are correct again, was a bug, caused by the new `enzymesCustom` prop. I fixed in the latest version `3.0.31`. Example below where I just used `enzymesCustom`:

```
const viewerInput = {
  // name: "L09136",
  // seq: "tcgcgcgtttcggtgatgacggtgaaaacctctgacacatgca",
  seq: "TTATGTATGCGTTGTCCTTGGAGTATTAATATTGTTCATGTGGGCAGGCTCAGGTTGAGGTTGAGGTTGAGGGAACTGCTGTTCCTGT",
  enzymesCustom: {
    Cas9: {
      rseq: "NGG", // recognition sequence
      fcut: 0, // cut index on FWD strand, relative to start of rseq
      rcut: 1, // cut index on REV strand, relative to start of rseq
    }
  },
  viewer: "both",
  // accession: "NC_011521",
  // backbone: "pSB1C3",
  showAnnotations: true,
  showPrimers: true,
  showComplement: true,
  showIndex: true,
  zoom: { linear: 50 },
  colors: [],
  onSearch: results => {
    console.log("Your Search Results: ", results);
  },
  bpColors: {
    10: "green",
    200: "blue",
    201: "red"
  },
  search: { query: "gtacc" },
  // enzymes: ["EcoRI"],
  copyEvent: event => event.key === "c" && (event.metaKey || event.ctrlKey),
  style: { height: "calc(100vh - 20px)", width: "calc(100vw)" }
};
```

<img width="677" alt="Screen Shot 2021-03-02 at 6 28 45 PM" src="https://user-images.githubusercontent.com/13923102/109729148-747abe00-7b85-11eb-8825-fcaf2e2e5452.png">

**Comment by jjti** [comment]
I'm going to move the feature request to another issue: https://github.com/Lattice-Automation/seqviz/issues/95
